### PR TITLE
WIP - Work In Progress (DO NOT MERGE): LCP and LSD

### PR DESCRIPTION
**Pull request by danielweck**
Fixes https://github.com/readium/readium-desktop/issues/663
Fixes https://github.com/readium/readium-desktop/issues/787
Fixes https://github.com/readium/readium-desktop/issues/326
Fixes https://github.com/readium/readium-desktop/issues/484
Fixes https://github.com/readium/readium-desktop/issues/325
Fixes https://github.com/readium/readium-desktop/issues/324
Fixes https://github.com/readium/readium-desktop/issues/706
Fixes https://github.com/readium/readium-desktop/issues/655

List of LCP-related issues:
https://github.com/readium/readium-desktop/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+lcp

### CODE REVIEW

**Comment by danielweck** [comment]
Current status:
* when importing a LCP license file into the library (e.g. `license.lcpl`): the referenced encrypted EPUB downloads fine, the `META-INF/license.lcpl` file entry is injected successfully into the ZIP archive, the resulting protected/licensed EPUB opens fine, but the passphrase / userkey fails (because I think that at this moment in time in the develop branch, the LCP-handling code has largely been disrupted / commented-out in order to perform a refactor on other critical parts of the codebase ... to be verified)
* When importing a protected/licensed EPUB directly (i.e. not via a `license.lcpl` file): there is a `publication.lsd` = `undefined` error in the console => this is "normal", due to incomplete LCP handling code.

**Comment by danielweck** [comment]
Updated status: LCP passphrase dialog reinstated with fixed lifecycle

File: src/renderer/redux/sagas/lcp.ts [hunk 1]
```diff
@@ -7,14 +7,31 @@
...
+// import { ReaderState } from "readium-desktop/renderer/redux/states/reader";
+
 export function* lcpUserKeyCheckRequestWatcher(): SagaIterator {
     while (true) {
         const action: any = yield take(lcpActions.ActionType.UserKeyCheckRequest);
 
         const { hint, publication, message } = action.payload;
+        // const id1 = (publication as PublicationView).identifier;
+
+        const store = diRendererGet("store");
```
Comment (panaC) [inline src/renderer/redux/sagas/lcp.ts]: Have you thought about that : https://redux-saga.js.org/docs/api/index.html#selectselector-args

Comment (danielweck) [reply src/renderer/redux/sagas/lcp.ts]: I would love to clean this up. Right now testing for a truthy condition of `rendererStore.getState().reader.reader` is totally hacky!
In fact, the "LCP passphrase" and "publication information" (etc.) dialogs do not trigger using the same technique, there are inconsistencies in the current implementation which make it hard to track the action flow (as a developer analysing the code).
Thanks for you tips, keep them coming! :)

**Comment by danielweck** [comment]
Hello @panaC , may I please ask for your review of this LCP / LSD Pull Request? I do not expect you to review the actual LCP/LSD logic (but feel free if you wish :), however I would like your input about the peripheral code changes.

Basically, I harmonized the `React.Component<iProps, iState>` so that the `mapDispatchToProps` and `mapStateToProps` do not have circular references (`ReturnType<typeof X>`). My initial motivation was to use consistent naming rules so that I could find my way around the codebase more easily (this helped me a lot to implement LCP/LSD support).

The problem is that this PR now contains lots of changes not directly related to LCP/LSD. I think that there are two options:

1) approve the PR as is, but be aware that existing pending PRs (especially your Redux/Saga/API refactorings, Pierre) will have to resolve conflicts (probably minor, but a little bit time consuming).
2) amend the PR to completely remove non-LCP/LSD changes. This will take a lot of time, but I am open to discuss options.

At any rate, LCP/LSD support should be merged into `develop` soon, because other developers / integrators depend on it.

**Pull request merged**